### PR TITLE
Implement langpacks API (bug 1122263)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1122263

- [x] `GET /langpacks/`
- [x] `GET /langpacks/<uuid>/`
- [x] `POST /langpacks/`
- [x] `PUT /langpacks/<uuid>/`
- [x] `PATCH /langpacks/<uuid>/`
- [x] `DELETE /langpacks/<uuid>/`
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by diox** [comment]
My idea for the API is to have:
- `GET /langpacks/` - listing available langpacks (with optional filters through query parameters)
- `GET /langpacks/<uuid>/` - list a single langpack (probably unused)
- `POST /langpack/` - upload a new langpack .zip file
- `PUT /langpacks/<uuid>/' - upload a langpack .zip file to replace a specific langpack
- `PATCH /langpacks/<uuid>/' - change the status of a specific langpack (all other fields are manifest-driven so you can't change them without uploading a new package)

Let me know if this sounds crazy, @robhudson, @chuckharmston ?

Note: I wanted to re-use the same principles governing the apps API but it turns out we don't have an endpoint to upload a new version for a `Webapp`, only an endpoint to upload a new `Webapp`. In addition, I didn't make a `Version` model for Langpacks, because this was overkill and uneeded at the moment, so I can't copy the apps API.

**Comment by chuckharmston** [comment]
Did you mean `POST /langpacks/`? I'm not crazy about the Rails-y pattern of switching between singular and plural in the URIs. Otherwise, I think the ideas are sound.

**Comment by diox** [comment]
Yeah `/langpacks/` all around, I fixed the comment

**Comment by chuckharmston** [comment]
I especially appreciate the creative (and appropriate!) uses of PUT/PATCH. Very nice :+1:

**Comment by diox** [comment]
This is now ready for review.

File: mkt/langpacks/views.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+from mkt.langpacks.models import LangPack
+from mkt.langpacks.serializers import (LangPackSerializer,
+                                       LangPackUploadSerializer)
+
+
+log = commonware.log.getLogger('z.api')
+
+
+
+# Flow is similar to apps (FIXME: add this to docs/):
```
Comment (robhudson) [inline mkt/langpacks/views.py]: This looks done already.

File: docs/api/topics/langpacks.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    :param objects: A :ref:`listing <objects-response-label>` of :ref:`langpacks <langpack-response-label>`.
+    :type objects: array
+
+Detail
+======
+
+.. _langpack-response-label:
+.. http:get:: /api/v2/langpacks/(string:uuid)/
+
+    Returns a single langpack. If the request is authenticated and the user has the ``LangPacks:%``, inactive langpacks
```
Comment (robhudson) [inline docs/api/topics/langpacks.rst]: 'permission' word is missing

File: docs/api/topics/langpacks.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    can be returned.
+
```
Comment (robhudson) [inline docs/api/topics/langpacks.rst]: Should this have the `**Response**` heading?

File: mkt/langpacks/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    def setUp(self):
+        super(TestLangPackViewSetMixin, self).setUp()
+        self.list_url = reverse('api-v2:langpack-list')
+        self.user = UserProfile.objects.get(pk=2519)
+
+    def create_langpack(self, **kwargs):
+        data = {
+            'filename': 'dummy.zip',
+            'hash': 'dummy-hash',
+            'size': 666,
```
Comment (robhudson) [inline mkt/langpacks/tests/test_views.py]: :smiling_imp:

File: mkt/langpacks/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+
+        response = self.client.get(self.list_url)
+        eq_(response.status_code, 200)
+        ok_(len(response.json['objects']), 1)
+        self.check_langpack(response.json['objects'][0])
+
+        response = self.client.get(self.list_url)
+        eq_(response.status_code, 200)
+        ok_(len(response.json['objects']), 1)
+        self.check_langpack(response.json['objects'][0])
```
Comment (robhudson) [inline mkt/langpacks/tests/test_views.py]: This block looks the same as the block above it. Curious if you intended the detail url there?

Comment (diox) [inline mkt/langpacks/tests/test_views.py]: Actually it's the same as the one below - the one above was anonymous. I splitted the test and removed the extra block.

File: mkt/langpacks/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+        eq_(response.json, {u'upload': [u'No upload found.']})
+
+    def test_dont_own_the_upload(self):
+        FileUpload.objects.create(uuid='my-uuid', user=None, valid=True)
+        self.grant_permission(self.user, 'LangPacks:%')
+
+        response = self.client.post(self.list_url, data=json.dumps({
+            'upload': 'my-uuid'}))
+        eq_(response.status_code, 400)
+        eq_(response.json, {u'upload': [u'No upload found.']})
```
Comment (robhudson) [inline mkt/langpacks/tests/test_views.py]: Good one. :+1:

File: mkt/langpacks/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+        FileUpload.objects.create(uuid='my-uuid', valid=False, user=self.user)
+        self.grant_permission(self.user, 'LangPacks:%')
+
+        response = self.client.post(self.list_url, data=json.dumps({
+            'upload': 'my-uuid'}))
+        eq_(response.status_code, 400)
+        eq_(response.json, {u'upload': [u'Upload not valid.']})
+
+    @patch('mkt.langpacks.models.LangPack.from_upload')
+    def test_errors_returned_by_from_upload(self, mock_from_upload):
```
Comment (robhudson) [inline mkt/langpacks/tests/test_views.py]: the `by_from` doesn't quite sound right?

Comment (diox) [inline mkt/langpacks/tests/test_views.py]: well, it's to test errors returned by `from_upload()`

Comment (robhudson) [inline mkt/langpacks/tests/test_views.py]: ah, I see it now :)

**Comment by robhudson** [comment]
Great work and awesome tests. r+wc.

**Pull request merged**